cranelift/codegen/src/ir/pcc.rs

```diff
@@ -257,135 +265,128 @@ impl BaseExpr {
         lhs == rhs || *lhs == BaseExpr::None || *rhs == BaseExpr::Max
     }
+}
 
-    /// Compute some BaseExpr that will be less than or equal to both
-    /// inputs. This is a generalization of `min` (but looser).
-    fn min(lhs: &BaseExpr, rhs: &BaseExpr) -> BaseExpr {
-        if lhs == rhs {
-            lhs.clone()
-        } else if *lhs == BaseExpr::Max {
-            rhs.clone()
-        } else if *rhs == BaseExpr::Max {
-            lhs.clone()
-        } else {
-            BaseExpr::None // zero is <= x for all (unsigned) x.
+impl Expr {
+    /// Constant value.
+    pub const fn constant(value: u64) -> Self {
+        Expr {
+            base: BaseExpr::None,
+            // Safety: `i128::from(u64)` is not const, but this will never overflow.
+            offset: value as i128,
         }
     }
 
-    /// Compute some BaseExpr that will be greater than or equal to
-    /// both inputs.
-    fn max(lhs: &BaseExpr, rhs: &BaseExpr) -> BaseExpr {
-        if lhs == rhs {
-            lhs.clone()
-        } else if *lhs == BaseExpr::None {
-            rhs.clone()
-        } else if *rhs == BaseExpr::None {
-            lhs.clone()
-        } else {
-            BaseExpr::Max
+    /// Constant value, full 128-bit width.
+    pub const fn constant128(value: i128) -> Self {
+        Expr {
+            base: BaseExpr::None,
+            offset: value,
         }
     }
-}
 
-impl Expr {
-    /// Constant value.
-    pub fn constant(offset: i64) -> Self {
+    /// Maximum (saturated) value.
+    pub const fn max_value() -> Self {
         Expr {
-            base: BaseExpr::None,
-            offset,
+            base: BaseExpr::Max,
+            offset: 0,
         }
     }
 
     /// The value of an SSA value.
-    pub fn value(value: ir::Value) -> Self {
+    pub const fn value(value: ir::Value) -> Self {
         Expr {
             base: BaseExpr::Value(value),
             offset: 0,
         }
     }
 
+    /// The value of an SSA value plus some offset.
+    pub const fn value_offset(value: ir::Value, offset: i128) -> Self {
+        Expr {
+            base: BaseExpr::Value(value),
+            offset,
+        }
+    }
+
     /// The value of a global value.
-    pub fn global_value(gv: ir::GlobalValue) -> Self {
+    pub const fn global_value(gv: ir::GlobalValue) -> Self {
         Expr {
             base: BaseExpr::GlobalValue(gv),
             offset: 0,
         }
     }
 
+    /// The value of a global value plus some offset.
+    pub const fn global_value_offset(gv: ir::GlobalValue, offset: i128) -> Self {
+        Expr {
+            base: BaseExpr::GlobalValue(gv),
+            offset,
+        }
+    }
+
     /// Is one expression definitely less than or equal to another?
     /// (We can't always know; in such cases, returns `false`.)
     fn le(lhs: &Expr, rhs: &Expr) -> bool {
-        if rhs.base == BaseExpr::Max {
+        let result = if rhs.base == BaseExpr::Max {
+            true
```

Comment:
What if `lhs.base == BaseExpr::Max` but `lhs.offset > rhs.offset`?

Reply:
Good point on this and below; reworked so there is no `BaseExpr` and `Expr` is now the enum with `Max` as a variant.